runtime-transaction/src/instructions_processor.rs

```diff
@@ -1,122 +1,34 @@
-                        return Err(duplicate_instruction_error);
-                    }
-                    if sanitize_requested_heap_size(bytes) {
-                        requested_heap_size = Some(bytes);
-                    } else {
-                        return Err(invalid_instruction_data_error);
-                    }
-                }
-                Ok(ComputeBudgetInstruction::SetComputeUnitLimit(compute_unit_limit)) => {
-                    if updated_compute_unit_limit.is_some() {
-                        return Err(duplicate_instruction_error);
-                    }
-                    updated_compute_unit_limit = Some(compute_unit_limit);
-                }
-                Ok(ComputeBudgetInstruction::SetComputeUnitPrice(micro_lamports)) => {
-                    if updated_compute_unit_price.is_some() {
-                        return Err(duplicate_instruction_error);
-                    }
-                    updated_compute_unit_price = Some(micro_lamports);
-                }
-                Ok(ComputeBudgetInstruction::SetLoadedAccountsDataSizeLimit(bytes)) => {
-                    if updated_loaded_accounts_data_size_limit.is_some() {
-                        return Err(duplicate_instruction_error);
-                    }
-                    updated_loaded_accounts_data_size_limit = Some(
-                        NonZeroU32::new(bytes)
-                            .ok_or(TransactionError::InvalidLoadedAccountsDataSizeLimit)?,
-                    );
-                }
-                _ => return Err(invalid_instruction_data_error),
-            }
-        } else {
-            // only include non-request instructions in default max calc
-            num_non_compute_budget_instructions =
-                num_non_compute_budget_instructions.saturating_add(1);
-        }
-    }
-
-    // sanitize limits
-    let updated_heap_bytes = requested_heap_size
-        .unwrap_or(MIN_HEAP_FRAME_BYTES) // loader's default heap_size
-        .min(MAX_HEAP_FRAME_BYTES);
-
-    let compute_unit_limit = updated_compute_unit_limit
-        .unwrap_or_else(|| {
-            num_non_compute_budget_instructions
-                .saturating_mul(DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT)
-        })
-        .min(MAX_COMPUTE_UNIT_LIMIT);
-
-    let compute_unit_price = updated_compute_unit_price.unwrap_or(0);
-
-    let loaded_accounts_bytes = updated_loaded_accounts_data_size_limit
-        .unwrap_or(MAX_LOADED_ACCOUNTS_DATA_SIZE_BYTES)
-        .min(MAX_LOADED_ACCOUNTS_DATA_SIZE_BYTES);
-
-    Ok(ComputeBudgetLimits {
-        updated_heap_bytes,
-        compute_unit_limit,
-        compute_unit_price,
-        loaded_accounts_bytes,
-    })
-}
-
-fn sanitize_requested_heap_size(bytes: u32) -> bool {
-    (MIN_HEAP_FRAME_BYTES..=MAX_HEAP_FRAME_BYTES).contains(&bytes) && bytes % 1024 == 0
+) -> Result<ComputeBudgetLimits> {
+    InstructionDetails::try_from(instructions)?.sanitize_and_convert_to_compute_budget_limits()
```

Comment:
how much slower is this with the additional work to check against builtins?

Reply:
Additional cost is a lookup at AhashMap of BUILTIN. I'll bench it. But this additional check against builtin cost is necessary to fix #682

Reply:
I made a comment there - I don't think we even should "fix" that.

Reply:
I think eventually it makes sense that we have this; but in the currenet state there are many calls where we are only getting the builtin or only the compute details.

We shouldn't require getting both since its' not necessary. We don't want to make the current situation worse.
I think we should probably (for now) allow individual processing of getting compute-budget so we don't make the current (numerous) calls worse.

Reply:
sorry, I don't think I follow, can you explain your concern (eg "make current situation worse") more? 

It's probably my mistake to bring up #682 here, tho I think it's relevant. (let's discuss that in that issue itself).

Another issue trying to solve is #2212, where compute-budget allocates 200K budget per builtin instruction. That behavior is incorrect, to address it, it needs to lookup BUILTINs AhashMap.

Reply:
Gotcha. #2212 seems more relevant to me.

Currently, in a bunch of places we are re-computing the compute-budget-limits.
Right now, we simply do a single pubkey == check to look for compute-budget. If compute-budget do some extra processing.

With the code in this PR, we are now also going to do a pubkey hash per program key to look for builtins even when we don't need them (once #2212 activates we will need them).

A single hash doesn't seem bad, but its actually pretty expensive in the degenerate cases where txs have many instructions.
It's still probably not going to be a major bottleneck

Reply:
Ah, thanks for explaining, I see you point. Where this goes is runtime-transaction call InstructionDetails::try_from() once, then cache result; eliminating repeated calls to iterating instructions. We should expecting saving overall. 

Unrelated, but solana-builtin-default-costs crate also has a lot room for improving, as we talked about before, the first thing needs to address there is it should have cost per instruction, not cost per program-id. That'd change pubkey hashing discussion.

Reply:
Yeah I 100% agree that once we start using runtime transaction it will save overall. The issue I am pointing out is that we currently don't use it, so this is making our repeated calls worse